src/v/datalake/record_multiplexer.cc

```diff
@@ -34,11 +46,61 @@ record_multiplexer::operator()(model::record_batch batch) {
         auto record = it.next();
         iobuf key = record.release_key();
         iobuf val = record.release_value();
-        // *1000: Redpanda timestamps are milliseconds. Iceberg uses
-        // microseconds.
-        int64_t timestamp = (first_timestamp + record.timestamp_delta()) * 1000;
+        auto timestamp = model::timestamp{
+          first_timestamp + record.timestamp_delta()};
         kafka::offset offset{batch.base_offset()() + record.offset_delta()};
-        int64_t estimated_size = key.size_bytes() + val.size_bytes() + 16;
+        int64_t estimated_size = key.size_bytes() + val.size_bytes();
+
+        auto val_type_res = co_await _type_resolver.resolve_buf_type(
+          std::move(val));
+        if (val_type_res.has_error()) {
+            switch (val_type_res.error()) {
+            case type_resolver::errc::bad_input:
+            case type_resolver::errc::translation_error:
+                // TODO: metric + use binary data?
+            case type_resolver::errc::registry_error:
+                _error = data_writer_error::parquet_conversion_error;
+                co_return ss::stop_iteration::yes;
+            }
+        }
+
+        auto record_data_res = co_await record_translator::translate_data(
+          offset,
+          std::move(key),
+          val_type_res.value().type,
+          std::move(val_type_res.value().parsable_buf),
+          timestamp);
+        if (record_data_res.has_error()) {
+            switch (record_data_res.error()) {
+            case record_translator::errc::translation_error:
+                // TODO: metric + use binary data?
```

Comment:
Does this mean a malformed non conforming value blocks translation from progressing, it should hit the same condition everytime IIUC.

Reply:
That's right. I opted to keep this case simple for the sake of introducing translation. I think we need to settle on the policy for bad data (long term, a DLQ) which I think will likely be to drop the data for the beta.

Reply:
isn't it better to add them as raw binaries instead of dropping?

Reply:
Yea that was my gut too, but it seems there's a class of user who want the queryable Iceberg data to only include valid data. This is an active topic in the ongoing discussion on schema compatibility

Reply:
There are 4 options:

1. Drop data
2. Write to another topic
3. Write to the same iceberg table
4. Write to another iceberg table

My doc in #p-iceberg goes over the options and what I recommend. Just need to convince Matt. I think writing to another topic gets really tricky with delivery guarantees. Seems simpler to keep it in the iceberg domain (which maybe means another iceberg table).

Reply:
cool, I missed the doc, went thru it just now. It makes sense.. I was also thinking if we can exploit any partitioning schemes in iceberg and use a separate hidden partition to store DLQ records so we don't have yet another iceberg table to write to.. but AFAICT there is no way to filter that partition out for all scans, particularly when there are no predicates.